tools/misc/check-notebooks.js

```diff
@@ -0,0 +1,137 @@
+};
+let _or = (...expecteds) => (value, path) => {
+  let errors = expecteds.map(expected => {
+    let e = null;
+    try { _verifier(expected)(value, path); return null; }
+    catch (err) { if (!(err instanceof CheckError)) throw err;
+                  else return err; }
+  }).filter(x => x !== null)
+  if (errors.length == expecteds.length)
+    bad("OR(" + expecteds.length + "):\n" +
+        errors.map(e => (e.message)).join("\n"));
+}
+
+let _verifier = expected => {
+  switch (type(expected)) {
+    case "function": return expected;
+    case "array":    return _arr(expected)
+    case "object":   return _obj(expected);
+    default:         return _eq(expected);
+  }
+}
+
+function check(file, expected) {
+  process.stdout.write(`Checking ${path.basename(file)}... `);
+  let str = fs.readFileSync(file, "utf8");
+  let x = JSON.parse(str);
+  { let nstr = normalString(x);
+    if (str != nstr) {
+      fs.writeFileSync("/tmp/check.json", nstr);
+      bad("not normalized (indentation?)\n  compare with \"/tmp/check.json\"");
+    }
+  }
+  _verifier(expected)(x, "TOP");
+  process.stdout.write(`OK\n`);
+}
+
+function checkAll(files, expected) {
+  try { files.forEach(file => check(file, expected)); }
+  catch (err) {
+    if (!(err instanceof CheckError)) throw err;
+    process.stdout.write(`FAIL!\n${err.message}\n`);
+  }
+}
+
+let _cellMetadata =
+  _or({},
+      {"mml-deploy": "local",     collapsed: false},
+      {"mml-deploy": "hdinsight", collapsed: true});
+let _srcStrings = _arrOf((value, path) => {
+  assertType(value, "string", path);
+  if (value.endsWith(" ")) bad(`${path}: bad string, ends with a space`);
+  if (value.length > 120)  bad(`${path}: bad string, too long`);
+  if (value.match(/\n./))  bad(`${path}: bad string, \\n not and EOS`);
+});
+
+let getNotebooks = () => {
+  let rx = /\.ipynb$/;
+  let scanDir = (dir) => {
+    let paths = fs.readdirSync(dir).map(p => path.join(dir,p));
+    let dirs  = paths.filter(p => fs.statSync(p).isDirectory());
+    let files = paths.filter(p => dirs.indexOf(p) < 0);
+    return files.filter(f => rx.test(f)).concat(...(dirs.map(scanDir)));
+  };
+  let baseDir = path.dirname(path.dirname(path.dirname(__filename)));
+  return scanDir(baseDir);
+};
+
+let notebooks =
+  (process.argv.length > 2) ? process.argv.slice(2) : getNotebooks();
+
+checkAll(
+  notebooks,
+  {cells:
+   _arrOf(_or({cell_type: "markdown", metadata: _cellMetadata,
+               source: _srcStrings},
+              {cell_type: "code", execution_count: null,
+               metadata: _cellMetadata, outputs: [],
+               source: _srcStrings})),
+   metadata: _obj({"anaconda-cloud": _obj({}),
+                   kernelspec: _obj({display_name: "Python [default]",
+                                     language: "python",
+                                     name: "python3"}),
+                   language_info: _}),
+   nbformat: 4,
+   nbformat_minor: 0});
```

Comment:
Will this just block notebooks if they have funky fields? If so, is there a way to make it a script that normalizes it to take the tedium out of it?

Reply:
@mhamilton723, no -- it's a script that I'll use to verify and fix new notebooks when they come in, since I got tired of doing the fixes manually and missing things over and over again...  It'll also be useful in making changes to the normalized versions in case of changes -- eg, if @drdarshan thinks that `collapsed` should not be changed, then it'll be easy to do that over all notebooks quickly.

Reply:
We can just use nbformat to do all this. I'm not sure adding a dependency on node is really needed.